T&A 45372: Fixes inconsistent behavior in export/import incl. Participant Results

matheuszych opened the pull request:
https://mantis.ilias.de/view.php?id=45372

Aims to fix inconsistent behavior in export/import incl. Participant Results.

Already reviewed by @thojou.

matheuszych committed:
Improves participant data export and import and also fixes inconsistent participant naming in different tables

matheuszych committed:
Fixes unit tests

kergomard reviewed (changes requested):
Thank you very much for the PR @matheuszych !

As always I've my small nitpicks:
- [ ] Could you please make the conversion explicit [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-2b2e8009fe3ee31b614f56f8227ba6bb44b5305b43b5f08e86705705dcf34bc6R3595): `$main_settings->getGeneralSettings()->getAnonymity() ? '1' : '0'`.
- [ ] [This here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-a25e8bc63216ea2d9c6ba4e644b978a00881a7e73ef74a9bb60e4502f9d2c688R190) can and will go sideways: You are explicitly allowing the value to be null, but then you cannot pass it to `ilTestParticipantList::initializeFromDBRows(array $db_rows)` where the parameter is not nullable.
- [ ] A question: Why are we changing [this here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-b239abac27364f1a2c40ba1450e8d4588555e63842732f07dd522d18f591cfb0R365)? If the test is set to be anonymous, shouldn't we simply return "Anonymous" and be done with it?
- [ ] This is one for the personal taste (but not only): I do not like the assignment chains like [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-4c0c9752016e9033a1e229adb5d7ed7ded9fea043d88c11de4ad9cfe1a0c06d0R64) as it is very easy to miss that something has been assigned to two variables at once. Please remove `$active_id` and access `$row['active_id']` directly in the one single place we need it afterwards.

Thank you very much again and best,
@kergomard

matheuszych committed:
Adds improvements after review

matheuszych commented:
Hello @kergomard 

Thank you for your review!

- If ok i would prefer leaving this as is, because other boolean values are also converted this way later on in the method. This makes it cohere to the rest ([here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-2b2e8009fe3ee31b614f56f8227ba6bb44b5305b43b5f08e86705705dcf34bc6R3619), [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-2b2e8009fe3ee31b614f56f8227ba6bb44b5305b43b5f08e86705705dcf34bc6R3648) or [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-2b2e8009fe3ee31b614f56f8227ba6bb44b5305b43b5f08e86705705dcf34bc6R3673)) But if you prefer i will change it.

- I added an explicit check to prevent calling the method `ilTestParticipantList::initializeFromDBRows(array $db_rows)` in the first place if strictly not needed.

- We sort of could do that but there are two arguements for this approach IMHO. First just returning `Anonymous` or even `$this->lng->txt('anonymous')` would mean we would have a sort of hard coded value. By calling `ilObjTestAccess::_getParticipantData($participant->getActiveId())` we can rely on an already existing and used method to handle the participant name without introducing another custom exception (centralized logic). Also the `ilObjTestAccess::_getParticipantData($participant->getActiveId())` method does add a ` (Imported)` suffix to the participants name in case that user was imported. This might be useful for some to differentiate between regular anonymous participants and imported anonymous participants. 

- I removed the `$active_id` variable and replaced both usages of it with `$row['active_id']`.

If you need me to change the first or third point, please let me know.

Best regards
@matheuszych

kergomard commented:
Hi @matheuszych 

Thank you very much for  adapting this!

> If ok i would prefer leaving this as is, because other boolean values are also converted this way later on in the method. This makes it cohere to the rest ([here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-2b2e8009fe3ee31b614f56f8227ba6bb44b5305b43b5f08e86705705dcf34bc6R3619), [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-2b2e8009fe3ee31b614f56f8227ba6bb44b5305b43b5f08e86705705dcf34bc6R3648) or [here](https://github.com/ILIAS-eLearning/ILIAS/pull/10233/files#diff-2b2e8009fe3ee31b614f56f8227ba6bb44b5305b43b5f08e86705705dcf34bc6R3673)) But if you prefer i will change it.

Please change all four of them.

> We sort of could do that but there are two arguements for this approach IMHO. First just returning Anonymous or even $this->lng->txt('anonymous') would mean we would have a sort of hard coded value. By calling ilObjTestAccess::_getParticipantData($participant->getActiveId()) we can rely on an already existing and used method to handle the participant name without introducing another custom exception (centralized logic). Also the ilObjTestAccess::_getParticipantData($participant->getActiveId()) method does add a  (Imported) suffix to the participants name in case that user was imported. This might be useful for some to differentiate between regular anonymous participants and imported anonymous participants.

Yes, on some level you are right, but there are huge "buts" there:
- The function `_getParticipantData()` does not do, what it advertises and additionally it is static. So we should actually get rid of it.
- We incur a penalty of a database roundtrip for basically nothing.
- I do not believe the behavior to be right after this change: If it is anonymous it should show "Anonymous", if you want it to show "Anonymous (imported)" (which is also fine with me), this wouldn't happen as far as I can see, as the 'fullname' that will be used as 'importname' would be empty.

I'm perfectly fine, if you want to remove `_getParticipantData()` and replace all usages with something better for trunk ;-) .

Thanks again and best,
@kergomard

matheuszych committed:
Adds more improvements resulting from review

matheuszych commented:
Hello @kergomard ,

i changed all boolean values to use the `? 1 : 0`. There were a couple more of them 😃 
Also replaced `_getParticipantData()` with `$this->lng->txt('anonymous')`.

Best regards
@matheuszych

kergomard commented:
Thank you very much for the Updates @matheuszych !

I merged this and then got stuck on the cherry-pick to ILIAS 10 as the whole manual scoring has been redone, ...and the same will happen again for trunk. Could you please provide PRs for ILIAS 10 and trunk? Otherwise I will never get around to the review of the last feature PR ;-).

Sorry and thanks!
@kergomard
